controllers/openstackdeploy_controller.go

```diff
@@ -0,0 +1,241 @@
+
+// +kubebuilder:rbac:groups=osp-director.openstack.org,resources=openstackdeploys,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=osp-director.openstack.org,resources=openstackdeploys/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=osp-director.openstack.org,resources=openstackdeploys/finalizers,verbs=update
+// +kubebuilder:rbac:groups=batch,resources=jobs,verbs=get;list;create;update;delete;watch
+
+// Reconcile - OpenStackDeploy
+func (r *OpenStackDeployReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = log.FromContext(ctx)
+
+	// Fetch the ProvisionServer instance
+	instance := &ospdirectorv1beta1.OpenStackDeploy{}
+	err := r.Client.Get(context.TODO(), req.NamespacedName, instance)
+	if err != nil {
+		if k8s_errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers. Return and don't requeue.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	//
+	// initialize condition
+	//
+	cond := &ospdirectorv1beta1.Condition{}
+
+	//
+	// Used in comparisons below to determine whether a status update is actually needed
+	//
+	currentStatus := instance.Status.DeepCopy()
+	statusChanged := func() bool {
+		return !equality.Semantic.DeepEqual(
+			r.getNormalizedStatus(&instance.Status),
+			r.getNormalizedStatus(currentStatus),
+		)
+	}
+
+	defer func(cond *ospdirectorv1beta1.Condition) {
+		//
+		// Update object conditions
+		//
+		instance.Status.CurrentState = ospdirectorv1beta1.DeployState(cond.Type)
+		instance.Status.CurrentReason = ospdirectorv1beta1.DeployReason(cond.Message)
+
+		instance.Status.Conditions.UpdateCurrentCondition(
+			cond.Type,
+			cond.Reason,
+			cond.Message,
+		)
+
+		if statusChanged() {
+			if updateErr := r.Client.Status().Update(context.Background(), instance); updateErr != nil {
+				if err == nil {
+					err = common.WrapErrorForObject(
+						"Update Status", instance, updateErr)
+				} else {
+					common.LogErrorForObject(r, updateErr, "Update status", instance)
+				}
+			} else {
+				// log status changed messages also to operator log
+				common.LogForObject(r, cond.Message, instance)
+			}
+		}
+
+	}(cond)
+
+	// look up the git secret from the configGenerator
+	configGenerator := &ospdirectorv1beta1.OpenStackConfigGenerator{}
+	err = r.GetClient().Get(context.TODO(), types.NamespacedName{Name: instance.Spec.ConfigGenerator, Namespace: instance.Namespace}, configGenerator)
+	if err != nil && errors.IsNotFound(err) {
+		r.GetLogger().Error(err, instance.Spec.ConfigGenerator+" ConfigGenerator not found!", "Instance.Namespace", instance.Namespace)
+	}
+
+	// Define a new Job object
+	job := openstackdeploy.DeployJob(instance, "openstackclient", instance.Spec.ConfigVersion, configGenerator.Spec.GitSecret)
```

Comment:
Is this used outside the scope of the if-block that starts on 148?  If not, we could move it inside to avoid calling the `DeployJob` func when the `ConfigVersion` matches between `Status` and `Spec`?

Reply:
good call, will move inside the if-block

Reply:
done in the latest commits